File: src/android/app/src/main/java/com/dogwalker/app/src/main/java/com/dogwalker/app/domain/usecase/walk/StartWalkUseCase.kt
package com.dogwalker.app.domain.usecase.walk

import com.dogwalker.app.data.repository.WalkRepository
import com.dogwalker.app.domain.model.Walk
import com.dogwalker.app.domain.model.Location
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.UUID

/**
 * Human Tasks:
 * 1. Verify that the database schema supports the Walk entity structure
 * 2. Ensure proper error handling is implemented in the UI layer for walk initiation failures
 * 3. Configure location services permissions in the Android Manifest
 * 4. Set up proper monitoring for walk initiation success/failure rates
 */

/**
 * Use case class that handles the business logic for initiating a new walk.
 * 
 * Requirements addressed:
 * - Service Execution (1.3 Scope/Core Features/Service Execution)
 *   Implements the core functionality for starting a walk session, including
 *   creating the walk entity, persisting it, and initializing GPS tracking.
 */
class StartWalkUseCase(
    private val walkRepository: WalkRepository
) {
    /**
     * Initiates a new walk by creating a Walk entity and persisting it in the database.
     * 
     * Requirements addressed:
     * - Service Execution (1.3 Scope/Core Features/Service Execution)
     *   Handles the initiation of a walk, including creating a Walk entity,
     *   persisting it, and starting live GPS tracking.
     *
     * @param userId ID of the user (walker) initiating the walk
     * @param dogId ID of the dog being walked
     * @param bookingId ID of the associated booking
     * @param initialLocation Initial GPS location where the walk starts
     * @return The newly created Walk entity
     * @throws IllegalArgumentException if any of the required parameters are invalid
     */
    suspend fun startWalk(
        userId: String,
        dogId: String,
        bookingId: String,
        initialLocation: Location
    ): Walk {
        // Validate input parameters
        require(userId.isNotBlank()) { "User ID cannot be blank" }
        require(dogId.isNotBlank()) { "Dog ID cannot be blank" }
        require(bookingId.isNotBlank()) { "Booking ID cannot be blank" }
        require(initialLocation.latitude in -90.0..90.0) { "Invalid latitude value" }
        require(initialLocation.longitude in -180.0..180.0) { "Invalid longitude value" }

        // Generate a unique ID for the new walk
        val walkId = UUID.randomUUID().toString()

        // Get current timestamp in ISO 8601 format
        val currentTime = DateTimeFormatter.ISO_INSTANT.format(Instant.now())

        // Create the initial walk entity
        val walk = Walk(
            id = walkId,
            userId = userId,
            dogId = dogId,
            bookingId = bookingId,
            locations = listOf(initialLocation),
            startTime = currentTime,
            endTime = currentTime, // Will be updated when walk ends
            status = "in_progress"
        )

        // Persist the walk entity in the database
        walkRepository.insertWalk(walk)

        return walk
    }
}

File: src/ios/DogWalker/Domain/UseCases/Walk/StartWalkUseCase.swift
// Foundation framework - Latest version
import Foundation

// MARK: - Human Tasks
/*
Prerequisites:
1. Configure location services in Info.plist with proper usage descriptions
2. Set up proper error handling for location tracking failures
3. Review and implement proper data retention policies for walk data compliance
4. Ensure proper access control for walk initiation
*/

// Internal imports using relative paths
import "../../../Domain/Entities/Walk"
import "../../../Domain/Entities/User"
import "../../../Domain/Entities/Location"
import "../../../Data/Repositories/WalkRepository"
import "../../../Data/Repositories/LocationRepository"

/// StartWalkUseCase: Responsible for initiating a new dog walking session
/// Requirement: Service Execution (1.3 Scope/Core Features/Service Execution)
/// Supports live GPS tracking, route recording, and location-based features during dog walking sessions
class StartWalkUseCase {
    
    // MARK: - Properties
    
    private let walkRepository: WalkRepository
    private let locationRepository: LocationRepository
    
    // MARK: - Initialization
    
    /// Initializes the StartWalkUseCase with required repositories
    /// - Parameters:
    ///   - walkRepository: Repository for managing Walk entities
    ///   - locationRepository: Repository for managing Location entities
    init(walkRepository: WalkRepository, locationRepository: LocationRepository) {
        self.walkRepository = walkRepository
        self.locationRepository = locationRepository
    }
    
    // MARK: - Public Methods
    
    /// Initiates a new walk session
    /// - Parameters:
    ///   - walker: The user who will conduct the walk
    ///   - dogs: Array of dogs participating in the walk
    ///   - startLocation: The starting location of the walk
    /// - Returns: The newly created Walk entity
    /// - Throws: Error if walk creation fails
    func startWalk(walker: User, dogs: [Dog], startLocation: Location) throws -> Walk {
        // Generate a unique ID for the walk
        let walkId = UUID().uuidString
        
        // Get current date/time for start time
        let startTime = Date()
        
        // Create the walk entity with initial status
        let walk = Walk(
            id: walkId,
            walker: walker,
            dogs: dogs,
            booking: Booking(
                id: UUID().uuidString,
                owner: User(
                    id: "",
                    name: "",
                    email: "",
                    phone: "",
                    role: "",
                    walks: [],
                    payments: [],
                    currentLocation: Location(latitude: 0, longitude: 0)
                ),
                walker: walker,
                dogs: dogs,
                walk: WalkSummary(
                    id: walkId,
                    distance: 0,
                    startTime: startTime,
                    endTime: startTime.addingTimeInterval(3600)
                ),
                payment: PaymentSummary(
                    id: UUID().uuidString,
                    amount: 0,
                    currency: "USD"
                ),
                scheduledAt: startTime,
                status: "in_progress"
            ),
            route: [startLocation],
            startTime: startTime,
            endTime: startTime.addingTimeInterval(3600),
            status: "in_progress"
        )
        
        do {
            // Save the walk entity
            try walkRepository.createWalk(walk)
            
            // Save the starting location
            _ = try await locationRepository.updateLocation(startLocation)
            
            Logger.info("Successfully started walk with ID: \(walkId)")
            
            return walk
        } catch {
            Logger.error("Failed to start walk: \(error.localizedDescription)")
            throw error
        }
    }
}